### ISSUE TITLE
Request for feature: eliminate singly-implemented interfaces

### ISSUE THREAD

**Issue by bjitivo** [HaxeFoundation/haxe#4603]
In our large code base, a considerable amount of code can be eliminated from the runtime, with corresponding size benefits, if the haxe compiler could perform this optimization:

- After all types have been parsed, look for interfaces that have only one class implementing them.  For such an interface IA implemented by a single class A, replace all type occurrences of IA throughout the AST by A.

In other words, turn all references to "interfaces implemented by a single class" into "the single class".

These interfaces would then never be seen by the compiler back end code and could contribute nothing to the eventual application size.

Our application has ~3260 interfaces, of which ~1260 would be eliminated by this optimization.  Since interfaces have a code cost for all backends, and an especially onerous cost for the hxcpp backend, eliminating these interfaces would save us a good deal of binary size.

Furthermore, the cost of the dynamically bound call through an interface function is significant, and if such interfaces are eliminated and replaced by the class that implements them, then the function call themselves can be much more efficient, for backends such as hxcpp and hxjava.

**Comment by Simn** [comment]
This assumes that all relevant types are known at compile-time so it would have to be hidden behind a compiler flag in order to avoid problems with partial compilation/run-time type loading.

**Comment by back2dos** [comment]
First of all, is this not something the very last compilation step should do (i.e. linking in C++ and JIT for all other backends)? At any time before, you don't actually know that there is only one class.

Secondly, it is impossible in general. Example:

```
interface CanFoo {
   function foo():Void;
}

class SuperFoo<T> implements CanFoo {
   var value:T;
   public function new(value) this.value = value;
   public function foo() trace(value);
}

var foo:Foo = if (Math.random() > .5) new SuperFoo(1) else new SuperFoo("foo");
```

There is no suitable substitute for `Foo`.

If you take type parameters out of the equation, then it can be done. Not sure it should, but here's a macro-based POC: https://github.com/back2dos/single_impl

**Comment by bjitivo** [comment]
Agree with Simn, it would be a compiler option, not enabled by default.  I could see something like '-Delide_interfaces_except="regexp"', where a regular expression could be provided that would test against every interface name (with full package prefix) to allow some interfaces to excluded if there was a reason to do so.

I don't understand back2dos' first point.  The front end knows every type, doesn't it?  And it knows every part of the inheritence hierarchy of all classes and interfaces and it knows which classes implement which interfaces.  It must know these things in order to be able to type check.  So the front end is perfectly capable of eliding interfaces that it knows are only implemented by one class.   And I think the front end is the right place to do this as it would have to be done in duplicate identically by every back end.  Actually there are alot of things currently done in the back ends that should be done in the front end, but that's a different issue.

If there is only a single specialization of a class implementing an interface, then I would consider that to be "a single implementation of the interface", so I think the optimization could still be applied.  If the interface itself is parameterized, well ... I'm not sure what could be done.  I'd probably just punt on that one for now.

Thanks for the pointer to the proof of concept.  I will take a look.

**Comment by delahee** [comment]
+1 to this, I struggled with this topic too and would love a fix.

2015-10-23 19:37 GMT+02:00 bjitivo <notifications@github.com>:

> Agree with Simn, it would be a compiler option, not enabled by default. I
> could see something like '-Delide_interfaces_except="regexp"', where a
> regular expression could be provided that would test against every
> interface name (with full package prefix) to allow some interfaces to
> excluded if there was a reason to do so.
>
> I don't understand back2dos' first point. The front end knows every type,
> doesn't it? And it knows every part of the inheritence hierarchy of all
> classes and interfaces and it knows which classes implement which
> interfaces. It must know these things in order to be able to type check. So
> the front end is perfectly capable of eliding interfaces that it knows are
> only implemented by one class. And I think the front end is the right place
> to do this as it would have to be done in duplicate identically by every
> back end. Actually there are alot of things currently done in the back ends
> that should be done in the front end, but that's a different issue.
>
> If there is only a single specialization of a class implementing an
> interface, then I would consider that to be "a single implementation of the
> interface", so I think the optimization could still be applied. If the
> interface itself is parameterized, well ... I'm not sure what could be
> done. I'd probably just punt on that one for now.
>
> Thanks for the pointer to the proof of concept. I will take a look.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/issues/4603#issuecomment-150641862>
> .
>



-- 
David Elahee

**Comment by bjitivo** [comment]
back2dos, I think your proof of concept implementation is missing something.

A class that implements an interface should be counted as also implementing all interfaces that that interface extends.  I believe this would be handled by having the onGenerate logic in FirstPass.hx not only push the class name onto the list of classes implementing the interface, but also push the class name into the list of implementing classes for all interfaces that that interface extends.

**Comment by ncannasse** [comment]
That seems to be a bit too particular optimization to go into the compiler.

Would you consider instead the following solution ?
a) detect these "lonely interfaces" by macros
b) generate stubs  `typedef MyLonelyInterface = MyActualClass` in a special directory
c) when compiling your "release builds", adds a `-cp mySpecialDirectory` so it uses the typedefs instead of the interfaces

**Comment by bjitivo** [comment]
I propose that this optimization is simple enough to implement in the compiler and broadly useful enough (interfaces with a single implementation in a given application are common because of design methodologies that advise creating interfaces to allow future expandability in areas that are not yet expanded) that it makes sense to go in the front end.  The back ends will not even have to process interface definitions for interfaces which are completely superfluous for the program being compiled.

**Comment by ncannasse** [comment]
@bjitivo the problem is that it only works if there's no runtime code loading, which is common on some platforms and might also be used on CPP in the future. We cannot really enable it by default, and having to activate it based on a EReg makes it quite difficult.

Another (maybe easier) option : detect the interfaces than mark them as extern + @:native("ActualClassName") with a macros ? (not sure it it works for CPP given that interface and class access might generate different code).

**Comment by bjitivo** [comment]
I don't understand how runtime code loading enters into this.  Are you saying that it's possible to load new types at runtime?  How does that work?

**Comment by ncannasse** [comment]
@bjitivo for instance on Flash you can load a SWF at runtime, this might be possible on some other platforms as well (C#/Java)

**Comment by bjitivo** [comment]
If it's not possible on all platforms, can you really support it as a language feature?

**Comment by ncannasse** [comment]
@bjitivo that's a bit O.T., that was just an example to illustrate that we don't want to enable this by default, since it might break on some platforms which allows runtime code loading/linking.

**Comment by bjitivo** [comment]
I don't understand how runtime code loading/linking that introduces new types can work across all targets.  In which case, I don't understand how it could be considered a possible future language feature.  In which case I don't understand how it's an example of something that "could be broken" by this optimization.  Something that will never happen cannot be broken.

**Comment by bjitivo** [comment]
DCE in general would have similar problems in an environment in which code could be dynamically added at runtime.  Code at runtime may want to call function X which no code at compile time needed, but some dynamically loaded code at runtime may need.  So your argument can be applied to DCE as well, which means that DCE should be removed as a feature from haxe because it might someday break runtime code loading.

Are you seeing yet how the argument doesn't make sense?

Anyway, removing lonely interfaces is just another form of DCE.  You've already accepted DCE into the haxe compiler, why not accept a proposal for improving it?

**Comment by waneck** [comment]
I agree that it's a pretty valid proposal, and it shouldn't be too hard to
implement.

2015-11-04 14:45 GMT-05:00 bjitivo <notifications@github.com>:

**Comment by bjitivo** [comment]
The macro supplied by back2dos doesn't work.

I'd really be interested in a compiler-based solution, which treats removing references to lonely interfaces as another aspect of DCE.  Making a macro to do this is really difficult.

**Comment by Simn** [comment]
There are problems when externs are involved in cases like this:

```haxe
interface I { }

class C implements I { }

extern class SomeApi {
	static public function getI():I;
}

class Main {
	static function main() {
		var i:I = SomeApi.getI();
	}
}
```

We would replace all occurrences of `I` with `C`, but that doesn't change the fact that `SomeApi.getI()` natively returns `I`. This is then likely to fail during native compilation.

I'm not sure if that's a big deal, but we should add a metadata which prevents interfaces from being eliminated so users can deal with these cases.

**Comment by bjitivo** [comment]
Can @:keep be applied to interfaces for this purpose?

**Comment by bjitivo** [comment]
Alternately ,it seems sensible to not remove interfaces for interfaces passed into or out of extern functions.

**Comment by Simn** [comment]
Maybe it's good enough to just ignore extern interfaces. It seems weird to have an extern API that returns non-extern interfaces.

**Comment by Simn** [comment]
There's another issue for our `@:overload` targets:

```haxe
interface I { }

class C implements I { }

@:keep
class Main {
	static function main() { }

	@:overload static public function foo(i:I) { }
	@:overload static public function foo(i:C) { }
}
```

As you can imagine replacing `I` with `C` leads to two `foo` methods with an identical signature, which causes a native compilation error.

**Comment by Simn** [comment]
Anyway, I've implemented this on this branch: https://github.com/HaxeFoundation/haxe/tree/issue_4603

It works with our unit tests but beyond that it's quite awkward to test properly. Could you try it on your codebase and report the results?

**Comment by bjitivo** [comment]
Are your changes supposed to remove the interface?  Because they don't seem to remove any interfaces from the compilation output.  It's important to remove entirely any interface that is not implemented at all any more after this feature is invoked.  Otherwise, the C++ backend emits the .h and .cpp files anyway and compiles them in and we don't save anything.

**Comment by waneck** [comment]
` let do_the_tivo_thing ctx =`
I lol'd

**Comment by bjitivo** [comment]
In addition, you changes seem to leave some constructs in the generated C++ code for the should-have-been-removed lonely interfaces.  Here is a sample program:

<pre>
class Main
{
    public static function main()
    {
        var i : ILonelyInterface = new LonelyImplementation();

        i.doSomething();
    }
}


private interface ILonelyInterface
{
    function doSomething() : Void;
}


private class LonelyImplementation implements ILonelyInterface
{
    public function doSomething() : Void
    {
        trace("Hello, world!");
    }

    public function new()
    {
    }
}
</pre>

Build with:

haxe -cpp out -main Main

The resulting out directory contains the generated C++ headers and source files.  Note that out/src/_Main/ILonelyInterface.h and out/src/_Main/ILonelyInterface.cpp are in there even though they shouldn't be.

Note also this line in out/include/_Main/LonelyImplementation.h:

<code>
       operator ::_Main::ILonelyInterface_obj *();
</code>

The generated lonely class apparently still knows something about being cast to the lonely interface type, but that interface type should have been removed.

In addition, when I build our actual sources, I get an error at one point due to a "CanCast" function being used with one of the "lonely interfaces", I don't understand the error entirely or the situation which reproduces it, but I will try to come up with a succinct test case the demonstrates the problem.

**Comment by bjitivo** [comment]
Looking at the change itself, I don't understand the code very well (because my brain basically rejects ocaml outright), but I don't see anywhere that would remove the 'implements' relationship between the lonely interface and its implementing class.  That's a fundamental aspect of the feature that appears to be missing.

Also, there's clearly nowhere that is removing the lonely interface from the AST representation entirely, which is also a missing piece of functionality.

**Comment by Simn** [comment]
You're right, the `implements` part was missing, try now.

**Comment by Simn** [comment]
Something else to consider:

```haxe
@:keep
interface I0 { }

interface I1 extends I0 { }

class InterfaceTest implements I1 {
	public function new() { }
}

class Main {
	static function main() {
		var itest = new InterfaceTest();
		var i0:I0 = itest;
	}
}
```

We can determine that `I1` can be substituted with `InterfaceTest`, but this loses the transitive relationship between `InterfaceTest` and `I0`. For now I'm simply not eliminating interfaces that extend other interfaces.

**Comment by back2dos** [comment]
A few words of caution.

1. We cannot just add this to DCE. If this behavior were introduced over night, we risk breaking all haxe/flash apps that load together multiple swfs, that are merely tied together by interfaces.
2. If this is about performance and the indirection introduced by an interface leads to a bottleneck, then this is the solution:

  ```haxe
  interface Dependency {
    function doTheOtherTiVoThing():Void;
  }
  class Highlander implements Dependency {
    function new() {}
    public function doTheOtherTiVoThing():Void trace('Here we are! Born to be kings ...');
    static public var THE_ONLY_ONE(default, null):Highlander;
  }
  class SlowedDown {
    public function new(d:Dependency) {
      d.doTheOtherTiVoThing();// <--- expensive indirection !!!
    }
  }
  @:generic 
  class NotSlowedDown<D:Dependency> {
    public function new(d:D) {
      d.doTheOtherTiVoThing();// <--- not any more !!!
    }
  }
  ```
 
 Note that the price for the indirection is removed **regardless of how many implementations there are**. I was somewhat disappointed to find that the compiler won't inline `doTheOtherTiVoThing` even if you tell it to (tried [here](http://try.haxe.org/#f4b2C)), but that's a different issue. The point is that it is possible to use this to remove the overhead of interfaces where it matters, and to do it even in cases when there are multiple implementations.
4. Instead of using interfaces, one can use abstracts to separate interface and implementation, resulting in the same abstraction barrier, but giving programmers full control when it comes to tweaking performance.

**Comment by Simn** [comment]
This will he hidden behind a compiler flag. This whole thing is about code output size which is a problem for TiVo.

**Comment by ncannasse** [comment]
@Simn would not a simpler solution would be to focus on supporting this only in genCPP : mark the the interfaces to be replaced with a metadata + extern, then when such a type is generated, substitute it by the class name stored in the metadata

**Comment by Simn** [comment]
You would have to ask @hughsando about the simplicity of that. I think my approach is more robust and versatile than a target-specific hack.

**Comment by waneck** [comment]
I agree. I can't see a target-specific solution being better in any sense

**Comment by hughsando** [comment]
Another possible optimization would be to find the interfaces that are only implemented by classes that do no inherit from other classes, and change the interface to class with empty functions.
This would allow multiple classes to implement the interface, but remove the indirection.
This would not work where you can't insert something at the top of the inheritance chain, but when it does work, it would give additional improvements. - possibly in combination with your other suggestion.
I have been thinking about multiple inheritance for interfaces.  This would not remove the indirection, but would transform it to a compiler-generated offset+jump, which is faster than re-pushing the params.
This would, however, make the implementation instances bigger as each would have to hold extra vtable pointers.  This still might work out well enough in the end.

**Comment by bjitivo** [comment]
Regarding not removing interfaces that extend other interfaces -- that is going to severely limit the value of this feature for us, since we have tons of such interfaces (our development team is a bit interface-happy, believing that it's a good design pattern to create an interface for pretty much every class, and this has led to many chains of interfaces each implemented by only a single class).

I don't quite understand the problem in Simn's proposed issue.  So I1 goes away and the class InterfaceTest no longer implements it.  Is the problem that this code will no longer work:

var foo = new InterfaceTest();

if (Std.is(foo, I0)) {
   // test succeeded
}

So this will no longer succeed because InterfaceTest is no longer an implementor of I0?

Why not just have the compiler replace the "implements I1" on InterfaceTest with "implements I0"?  In other words, when an interface is eliminated, if that interface extends an interface that is *not* eliminated, then the removed interface is replaced everywhere with its not-eliminated super-interface.

I realize that there are some serious gymnastics going on here, but it's an incredibly useful feature for us.

**Comment by bjitivo** [comment]
Thank you for the suggestions back2dos; but one problem for us is that we have ~700,000 lines of Haxe code with an incredible complexity of interface relationships and it would be a monumental task to rewrite it using the tricks you prescribe.  We realize that implementing this feature is going to take man-days of compiler developer work, but this is balanced against man-months of effort in reworking our code base.  We appreciate the haxe compiler developers donating some of their time here to help save us *alot* of time and money in such a rework.

**Comment by Simn** [comment]
Yes that's pretty much the problem. In fact `var i0:I0 = itest;` should already not compile natively (and doesn't on Java/C#) because the types are technically unrelated after their "common" interface has been removed.

I also thought along the lines of what you're suggesting with regards to interface inheritance. However, we should first make sure that the feature works in its current state before improving it further.

**Comment by bjitivo** [comment]
The latest changes don't seem to do anything beneficial.  The C++ ILonelyInterface header and source file are still generated.

I tried the "-xml" option in the compiler to see if it could elucidate anything.  It does show this:

    <class path="_Main.ILonelyInterface" params="" file="Main.hx" private="1" module="Main" interface="1">
        <doSomething public="1" set="method"><f a=""><x path="Void"/></f></doSomething>
        <meta><m n=":remove"/></meta>
    </class>

So there is a ":remove" meta applied to the interface, but that doesn't seem to be having the desired effect.  Is that something that the backend is supposed to act on?  Is the hxcpp backend not honoring this meta?

**Comment by Simn** [comment]
Are you testing with `-dce full`? Otherwise the compiler doesn't really touch your code.

**Comment by bjitivo** [comment]
Oh.  That's going to be a problem.  We can't use DCE on our code, we use too much reflection, and even when I reworked the code to "fix" the issues with reflect, I still found that some interface properties were getting eliminated in unexpected places and at that point gave up.

**Comment by bjitivo** [comment]
Also, I just tried it with -dce full and ILonelyInterface was still there.

Also, respectfully, is there a reason that you can't run the test that I provided above yourself to see whether or not you've fixed the issue? :)

**Comment by Simn** [comment]
I did and it's not there when compiling with `-dce full`. Did you clear the output directory before compiling again? Might be you just saw the ILonelyInterface files from a previous compilation.

**Comment by bjitivo** [comment]
I did clear it.  Something weird must be going on.

OK, I guess that somehow I had modified the test case accidentally when I was trying some other stuff.  I replaced it with the code that I have above in this issue and now the interface *is* being eliminated, but only if I use the -dce full option.

Hm, maybe I need to work harder on getting our code to be DCE-able.  Shame that I threw away all of my previous work when I hit the interface property wall ... that's going to take some time to reproduce.

**Comment by bjitivo** [comment]
p.s. sorry for the drama

**Comment by bjitivo** [comment]
OK so I think we're at a bit of an impasse; we need to be able to use DCE in our code base to be able to even utilize this feature, and that's going to take some significant work.  In the meantime, there are known issues with the feature as it's implemented (namely, not ever eliminating interfaces in chains even when it would be possible to do so, replacing them with super-interfaces where necessary) that could be worked on, but unless we're actually going to be able to successfully use DCE, that work may be wasted.

Just curious - why is it that the macro that back2dos wrote and that we modified to do the same thing as this compiler feature (but got blocked on the fact that Std.is() didn't have :genericBuild macros applied, although this has maybe been fixed?) was able to eliminate the interfaces even when we didn't run DCE -- but the compiler feature can't do that?

**Comment by Simn** [comment]
The macro completely ignores removable interfaces in its second pass. They are never even compiled, so there's no need to DCE them. This is obviously not possible to achieve in a single compilation pass.

I'm frankly surprised you're not using DCE given that code size is the issue at hand here. I'm not sure how to proceed with this. I could hack something into the branch that would DCE `@:remove` interfaces even in DCE mode, but that doesn't strike me as a worthwhile endeavor.

My favorite solution would be to just fix whatever DCE issue you came across, but I understand it requires some work to deal with reflection issues.

**Comment by bjitivo** [comment]
OK, we'll work on DCE.  For what it's worth, the size savings were not extreme even when I got our DCE'd build to mostly work.  We went from a 16 MB compressed binary to about 15 MB compressed.  That's sufficient to get over our current size hump should we get it to actually work, but it's not super exciting.  Also, given that the build didn't fully work because some needed stuff was still being removed, the 1 MB compressed size reduction was an upper bound on what we'd achieve by enabling DCE.  I expect that the total reduction would get cut in 1/2 to maybe 3/4 when all needed functionality was actually not DCEd, so maybe a 500 Kb to 750 Kb savings.  That's just on the edge of meeting our current needs; we need to reduce our compressed size by > 700 Kb ...

**Comment by Simn** [comment]
On a related note, I once made a pull request which dealt with reflection support code: https://github.com/HaxeFoundation/haxe/pull/3085

This never got merged into development but I could look into it again if you think it's worthwhile. However, given that you use a lot of reflection you might come across similar issues as with DCE.

**Comment by back2dos** [comment]
I have explored [a different and simpler option](https://github.com/back2dos/single_impl#second-approach-single_implsweep) that no longer requires two passes and does not require `@:genericBuild`, and is therefore unaffected by its limitations.

The whole logic that determines what is an orphan and what isn't has now been moved to `single_impl.Filter`.

I believe it now also deals with super-types of lonely interfaces and the new approach works with module subtypes also.

**Comment by bjitivo** [comment]
Thank you back2dos, your efforts are greatly and sincerely appreciated.

Does your new approach require "-dce full" to eliminate the interfaces from the generated code?

**Comment by Simn** [comment]
I considered using `@:native` but didn't like the idea of leaving the interfaces in the AST because I'm not sure how gencpp treats them, compared to classes. Even if the textual replacement works, you still operate on an interface as opposed to a class and that's not exactly the same thing.

**Comment by back2dos** [comment]
@bjitivo No, DCE is not required. I'm glad to help, although to be perfectly honest I do feel like I am just giving you a nicer rope to hang yourself with ... 700KLOC of code that is widely interspersed with reflection does sound like terribly high technical debt. But then again, that's for your team to deal with. I hope this makes it simpler.

@Simn If not generating the interface in order to reduce code size was the primary goal here (was it?), then I fail to see whether it matters that you operate on the interface itself. Or am I missing something? As for whether it works or not, there was only way to find out and it seems like I got lucky ;)

**Comment by bjitivo** [comment]
So far so good.  back2dos' latest macro is just killin' it on every small test I've tried so far.  Now the real challenge ... applying it to our actual code base.

**Comment by ncannasse** [comment]
@Simn can't you simply mark the removed interfaces as extern ?

**Comment by Simn** [comment]
That's what I'm doing now. I still don't like removing types outside DCE like that, but if it works for you guys I won't complain.

**Comment by ncannasse** [comment]
@hughsando any reason the extern interfaces are still generated in gencpp then?

**Comment by Simn** [comment]
They aren't, it's all good now.

**Comment by ncannasse** [comment]
@bjitivo could you confirm it's good for you?

**Comment by bjitivo** [comment]
I will do so.

**Comment by Simn** [comment]
Any news on this? Should we work towards getting this into the 3.3 release?
